### PR TITLE
feat: add request id to reporting

### PR DESCRIPTION
**Pull request by attriaayush**
- Adding `requestId` to error reporting. Follow pattern described here https://blog.sentry.io/2019/01/31/using-nginx-sentry-trace-errors-logs
- Moving analytics event reporting to once the retry mechanism is exhausted.

This should enable for us to correlate the events we see during error reporting and essentially thread it through the system.

### CODE REVIEW

File: src/snyk/snykCode/codeService.ts [hunk 1]
```diff
@@ -198,18 +198,18 @@ export class SnykCodeService extends AnalysisStatusProvider implements ISnykCode
       }
     } catch (err) {
       this.temporaryFailed = true;
-      await this.errorHandler.processError(err, undefined, () => {
+      await this.errorHandler.processError(err, { requestId }, () => {
         this.errorEncountered(requestId);
       });
 
-      if (enabledFeatures?.codeSecurityEnabled) {
+      if (enabledFeatures?.codeSecurityEnabled && this.errorHandler.transientErrorsLimitExhausted) {
```
Comment (bastiandoetsch) [inline src/snyk/snykCode/codeService.ts:205]: so we log it if `transientErrorsLimitExhausted` true? Either that is misnamed or I don't understand the logic.

File: src/snyk/snykCode/error/snykCodeErrorHandler.ts [hunk 2]
```diff
@@ -34,6 +36,10 @@ export class SnykCodeErrorHandler extends ErrorHandler implements ISnykCodeError
     this.transientErrors = 0;
   }
 
+  get transientErrorsLimitExhausted(): boolean {
+    return this.transientErrorLimitHit;
```
Comment (bastiandoetsch) [inline src/snyk/snykCode/error/snykCodeErrorHandler.ts:40]: Why the difference in naming?

File: src/snyk/snykCode/codeService.ts (hunk 1 above)
Comment (attriaayush) [reply src/snyk/snykCode/codeService.ts:205]: so the `transientErrorsLimitExhausted` correlates to the logic [here](https://github.com/snyk/vscode-extension/blob/d31f7d09d57c6b6efc6b79687a2642e65aa01139/src/snyk/snykCode/error/snykCodeErrorHandler.ts#L120). So essentially we retry failed request a bunch of times and check whether the amount of error'd tries are less than `MAX_CONNECTION_RETRIES`. Once that limit is exhausted or hit only then we should be logging an error event to amplitude since that will impact our customer's workflow. 

Reason for the naming was to match the naming of [`transientErrors`](https://github.com/snyk/vscode-extension/blob/d31f7d09d57c6b6efc6b79687a2642e65aa01139/src/snyk/snykCode/error/snykCodeErrorHandler.ts#L21) but happy to change it to `connectionRetryLimitExhausted` or something along those lines for better readability?

File: src/snyk/snykCode/error/snykCodeErrorHandler.ts (hunk 2 above)
Comment (attriaayush) [reply src/snyk/snykCode/error/snykCodeErrorHandler.ts:40]: just that `transientErrorsLimitExhausted` is a `public get` method on that class which reflects the value of a private variable `transientErrorLimitHit` which further records the `boolean` state of whether the limit was hit.

File: src/test/unit/base/services/authenticationService.test.ts [hunk 3]
```diff
@@ -154,6 +156,7 @@ suite('AuthenticationService', () => {
       {
         processError: sinon.fake(),
         resetTransientErrors: sinon.fake(),
+        transientErrorsLimitExhausted: false,
```
[the remark this answers is not in the document]
Comment (attriaayush) [reply src/test/unit/base/services/authenticationService.test.ts:159]: yeah had to update this file solely due to the breaking interface but this is an `authenticationService` and from what I understand we don't retry failed authentication requests.

File: src/snyk/snykCode/error/snykCodeErrorHandler.ts (hunk 2 above)
Comment (attriaayush) [reply src/snyk/snykCode/error/snykCodeErrorHandler.ts]: renamed the `transientErrorLimit` to `_transientErrorsLimitExhausted` to keep things consistent

Comment (attriaayush) [reply src/snyk/snykCode/error/snykCodeErrorHandler.ts]: renamed the `transientErrorLimit` to `_transientErrorsLimitExhausted` to keep things consistent

File: src/snyk/snykCode/error/snykCodeErrorHandler.ts [hunk 4]
```diff
@@ -132,6 +143,11 @@ export class SnykCodeErrorHandler extends ErrorHandler implements ISnykCodeError
       msg += `. ${options}`;
     }
 
+    // logging options to sentry as well
+    if (error instanceof Error && error) {
+      error.message = msg;
```
Comment (michelkaporin) [inline src/snyk/snykCode/error/snykCodeErrorHandler.ts:148]: Out of curiosity - how does `error.message` look here? An example would be nice to understand what Sentry will report 😄

File: src/test/unit/base/services/authenticationService.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (michelkaporin) [reply src/test/unit/base/services/authenticationService.test.ts:159]: Yeah, I think tests would be nice for `SnykCodeErrorHandler.transientErrorsLimitExhausted` as well. Ofc it should be part of `snykCodeErrorHandler.test.ts`.

[the remark this answers is not in the document]
Comment (attriaayush) [reply src/test/unit/base/services/authenticationService.test.ts:159]: yeah agree but my initial thoughts were to do something like an integration test for things like that specially when we have things like `setTimeout` in the game. 😄

File: src/snyk/snykCode/codeService.ts (hunk 1 above)
Comment (bastiandoetsch) [reply src/snyk/snykCode/codeService.ts:205]: Let's go with your connectionRetry suggestion.

File: src/test/unit/base/services/authenticationService.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bastiandoetsch) [reply src/test/unit/base/services/authenticationService.test.ts:159]: Let's just test reaction on Http error response, setting timeout to 1s

[the remark this answers is not in the document]
Comment (michelkaporin) [reply src/test/unit/base/services/authenticationService.test.ts:159]: Oh no, I wouldn't to any http requests, nor integration testing here. I'd just call a function and unit test if the class property has been changed. Let's keep things simple 😉

File: src/snyk/snykCode/error/snykCodeErrorHandler.ts (hunk 4 above)
Comment (attriaayush) [reply src/snyk/snykCode/error/snykCodeErrorHandler.ts]: Didn't really like this pattern tbf. 

Going with pattern defined here https://blog.sentry.io/2019/01/31/using-nginx-sentry-trace-errors-logs

Comment (attriaayush) [reply src/snyk/snykCode/error/snykCodeErrorHandler.ts]: Didn't really like this pattern tbf. 

Going with pattern defined here https://blog.sentry.io/2019/01/31/using-nginx-sentry-trace-errors-logs

File: src/snyk/common/error/errorReporter.ts [hunk 5]
```diff
@@ -78,6 +78,12 @@ export class ErrorReporter {
     }
   }
 
+  static configureScope(requestId: string) {
+    Sentry.configureScope(scope => {
```
Comment (michelkaporin) [inline src/snyk/common/error/errorReporter.ts:82]: This seems to be a global level configuration. Local scopes is probably a way to go here in order to operate on a single reported event.

File: src/test/unit/snykCode/error/snykCodeErrorHandler.test.ts [hunk 6]
```diff
@@ -42,4 +44,18 @@ suite('Snyk Code Error Handler', () => {
       }, CONNECTION_ERROR_RETRY_INTERVAL);
     });
   });
+
+  test('Logs analytic events once retries are exhausted', async function () {
```
Comment (michelkaporin) [inline src/test/unit/snykCode/error/snykCodeErrorHandler.test.ts:48]: 💟

**Review by michelkaporin — CHANGES_REQUESTED** [review]

File: src/snyk/common/error/errorReporter.ts (hunk 5 above)
Comment (attriaayush) [reply src/snyk/common/error/errorReporter.ts]: ah good shout.

**Review by bastiandoetsch — DISMISSED** [review]

File: src/snyk/common/error/errorReporter.ts [hunk 7]
```diff
@@ -10,6 +10,11 @@ import { OnUncaughtException } from './integrations/onUncaughtException';
 
 type SentryEnvironment = 'development' | 'preview' | 'production';
 
+export enum TagKeys {
+  CodeRequestId = 'code_request_id',
+}
+export type Tags = { [key in TagKeys]?: string };
```
Comment (michelkaporin) [inline src/snyk/common/error/errorReporter.ts:16]: It'd be cool to have types for values of the keys as well, but could be a pain to define them :) See if that's something feasible 😄 nice job anyway 👍

**Review by michelkaporin — COMMENTED** [review]
LGTM, let's fix the lint and then ✅

File: src/snyk/common/error/errorReporter.ts (hunk 7 above)
Comment (attriaayush) [reply src/snyk/common/error/errorReporter.ts:16]: Already discussed. 

tldr; Agree. We can always add types as we need to later on

**Review by michelkaporin — APPROVED** [review]
Sweet!
